tools/document-pii/src/pii_finder.rs

```diff
@@ -0,0 +1,281 @@
+    pub module_path: String,
+    pub current_type: String,
+    pub all_types: &'a HashMap<String, EnumOrStruct>,
+    pub use_statements: &'a BTreeMap<String, BTreeSet<String>>,
+    pub pii_values: &'a Vec<String>,
+    pub current_path: &'a mut Vec<TypeAndField>,
+    pub pii_types: &'a mut BTreeSet<Vec<TypeAndField>>, // output
+}
+
+/// Finds the full path to the given types by comparing them to the types in the scope.
+fn get_matching_use_paths<'a>(
+    field_types: &'a BTreeSet<String>,
+    local_paths: &'a BTreeSet<String>,
+) -> Vec<&'a String> {
+    local_paths
+        .iter()
+        .filter(|use_path| {
+            let last_use_path = use_path.split("::").last().unwrap().trim();
+            field_types
+                .iter()
+                .any(|field_type| field_type.trim() == last_use_path)
+        })
+        .collect()
+}
+
+impl<'a> PiiFinder<'a> {
+    /// Takes a Field and visit the types that it consist of if we have
+    /// the full path to it in self.use_statements.
+    fn visit_field_types(&mut self, ty: &Type) {
+        match ty {
+            Type::Path(path) => {
+                let local_paths = self.use_statements.get(&self.module_path).unwrap().clone();
+
+                let mut field_types = BTreeSet::new();
+                get_field_types(&path.path, &mut field_types);
+
+                let use_paths = get_matching_use_paths(&field_types, &local_paths);
+                for use_path in use_paths {
+                    if let Some(enum_or_struct) = self.all_types.get(use_path).cloned() {
+                        // Theses values will be changed when recursing, so we save them here so when we
+                        // return to this function after the match statement, we can set them back.
+                        let current_type = self.current_type.clone();
+                        let module_path = self.module_path.clone();
+                        self.module_path = use_path.rsplit_once("::").unwrap().0.to_owned();
+
+                        match enum_or_struct {
+                            EnumOrStruct::Struct(itemstruct) => self.visit_item_struct(&itemstruct),
+                            EnumOrStruct::Enum(itemenum) => self.visit_item_enum(&itemenum),
+                        }
+
+                        self.module_path = module_path;
+                        self.current_type = current_type;
+                    }
+                }
+            }
+            Type::Reference(reference) => self.visit_field_types(&reference.elem),
+            Type::Array(arr) => self.visit_field_types(&arr.elem),
+            Type::BareFn(bare_fn) => bare_fn
+                .inputs
+                .iter()
+                .for_each(|ty| self.visit_field_types(&ty.ty)),
+            Type::Group(group) => self.visit_field_types(&group.elem),
+            Type::ImplTrait(_) => {}
+            Type::Infer(_) => {}
+            Type::Macro(_) => {}
+            Type::Never(_) => {}
```

Comment:
can we combine all of the empty Types together at the end? 

```suggestion
            Type::Infer(_) | Type::Macro(_) | Type::Never(_) => {}
```